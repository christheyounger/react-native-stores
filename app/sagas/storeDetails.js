import { all, put, call, takeLatest } from 'redux-saga/effects'
import settings from "../config/settings";
import { handleErrors } from "../utilities";
import { 
    gotStore,
    getStoreFail,
    getStoreBranches,
    gotStoreBranches,
    getStoreBranchesFail
} from "../actions/stores";
import { GET_STORE, GET_STORE_BRANCHES } from '../constants/actiontypes';

export function* getStore(id) {
    try {
        const response = yield call(fetch, settings.apiUrl + 'store/' + id);
        handleErrors(response);
        const store = response.json();
        yield all([put(gotStore(store)), put(getStoreBranches(store.id))]);
    } catch (error) {
        yield put(getStoreFail(error));
    }
}

export function* watchGetStore() {
    yield takeLatest(GET_STORE, getStore);
}

export function* getStoreBranchesSaga(id) {
    try {
        const response = yield call(fetch, `${settings.apiUrl}'store/${id}/branches`)
        handleErrors(response);
        const branches = response.json();
        yield put(gotStoreBranches(branches));
    } catch (error) {
        yield put(getStoreBranchesFail(error));
    }
}

export function* watchGetBranches() {
    yield takeLatest(GET_STORE_BRANCHES, getStoreBranchesSaga);
}
