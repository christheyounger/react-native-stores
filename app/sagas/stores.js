import { put, call, takeLatest } from 'redux-saga/effects'
import settings from '../config/settings';
import { gotStores, getStoresFail } from '../actions/stores';
import { handleErrors } from '../utilities'

export function* getStores() {
    try {
        const response = yield call(fetch, settings.apiUrl + 'stores')
        handleErrors(response);
        const stores = response.json();
        yield put(gotStores(stores));
    } catch (error) {
        yield put(getStoresFail(error))
    }
}

export function* watchGetStores() {
    yield takeLatest('GET_STORES', getStores)
}