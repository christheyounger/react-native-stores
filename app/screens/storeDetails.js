import React from 'react';
import settings from '../config/settings';
import Store from '../components/Store';
import ConfirmModal from '../components/confirm';
import { ActivityIndicator, FlatList, TextInput, Text, View } from 'react-native';

export default class StoreDetail extends React.Component {
  static navigationOptions = {
    title: 'Store details',
  };

  constructor(props){
    super(props);
    this.state = {
      store: {name: props.store.name}
    };
  }

  componentDidMount(){
    this.props.getStore(this.storeId);
  }

  storeId = this.props.navigation.getParam('id', 'NO-ID');

  editName = name => {
    this.setState({store: {name}});
  }

  updateName = e => {
    const name = e.nativeEvent.text;
    fetch(`${settings.apiUrl}stores/${this.storeId}`,
      {
        method: 'PUT',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({name})
      })
    .then(response => response.json())
    .then(store => this.setState({store}))
  }

  handleClick(item) {
    this.props.navigation.push('StoreDetails', item);
  }

  delete() {
    fetch(`${settings.apiUrl}stores/${this.storeId}`, {method: 'DELETE'})
    .then(() => this.props.navigation.goBack());
  }

  render(){
    if (this.props.loading) {
      return(
        <View style={{flex: 1, padding: 20}}>
          <ActivityIndicator/>
        </View>
      );
    }

    return (
      <View style={{flex: 1, flexDirection: 'column', alignItems: 'center'}}>
        <TextInput
          onChangeText={name => this.editName(name)}
          onSubmitEditing={name => this.updateName(name)}
          style={{fontSize: 24}}
          value={this.state.store.name}
        />
        <View style={{flexDirection: 'row', height: 30, alignItems: 'stretch'}}>
          <ConfirmModal color='red' label='Delete' title={`Delete ${this.props.store.name}`} handleClick={() => this.delete()} />
        </View>
        <View>
          <Text style={{fontSize: 21}}>Branches</Text>
          <FlatList
            data={this.props.store.branches}
            renderItem={({item}) =>
              <Store
                id={item.id}
                name={item.name}
                branches={item.branches}
                handleClick={item => this.handleClick(item)}
              />}
            keyExtractor={({id}) => id.toString()}
          />
        </View>
      </View>
    )
  }
}
