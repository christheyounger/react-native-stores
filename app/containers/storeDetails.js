import { getStore, gotStore } from "../actions/stores";
import { connect } from 'react-redux';
import storeDetails from '../screens/storeDetails';

const matchStateToProps = (state) => {
    const { store: { loading, store} } = state;
    return { loading, store };
};

const mapDispatchToProps = dispatch => ({
    getStore: () => dispatch(getStore()),
});

export default connect(matchStateToProps, mapDispatchToProps)(storeDetails);
