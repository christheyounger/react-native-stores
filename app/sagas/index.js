import { all } from 'redux-saga/effects';
import { watchGetStores } from './stores';
import { watchGetStore } from './storeDetails';

export default function* rootSaga() {
    yield all([
        watchGetStores(),
        watchGetStore(),
    ])
}
