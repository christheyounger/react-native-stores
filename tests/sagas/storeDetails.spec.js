import 'isomorphic-fetch';
import { runSaga } from "redux-saga";
import { getStore, getStoreBranchesSaga } from '../../app/sagas/storeDetails';

describe('getStore', () => {
  const store = {id: 1, name: 'store 1'};
  test('should handle success', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };
    const mockPromise = Promise.resolve({
        ok: true,
        json: () => store,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    await runSaga({
        dispatch,
        getState: () => ({loading: false}),
    }, getStore);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
  test('should handle failure', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };
    const mockPromise = Promise.resolve({
        ok: false,
        statusText: 'not found',
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    await runSaga({
        dispatch,
        getState: () => ({loading: false}),
    }, getStore);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
})

describe('getStoreBranches', () => {
  const branches = [{id: 2, name: 'store 2'}];
  test('should handle success', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };
    const mockPromise = Promise.resolve({
        ok: true,
        json: () => branches,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    await runSaga({
        dispatch,
        getState: () => ({loading: false}),
    }, getStoreBranchesSaga);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
  test('should handle failure', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };
    const mockPromise = Promise.resolve({
        ok: false,
        statusText: 'not found',
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    await runSaga({
        dispatch,
        getState: () => ({loading: false}),
    }, getStoreBranchesSaga);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
})
