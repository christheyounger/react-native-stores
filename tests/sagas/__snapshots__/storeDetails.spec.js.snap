// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStore should handle failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/store/undefined",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "ok": false,
          "statusText": "not found",
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStore should handle failure 2`] = `
Array [
  Object {
    "payload": Object {
      "error": [Error: not found],
    },
    "type": "GET_STORE_FAIL",
  },
]
`;

exports[`getStore should handle success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/store/undefined",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "json": [Function],
          "ok": true,
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStore should handle success 2`] = `
Array [
  Object {
    "payload": Object {
      "store": Object {
        "id": 1,
        "name": "store 1",
      },
    },
    "type": "GOT_STORE",
  },
  Object {
    "payload": Object {
      "id": 1,
    },
    "type": "GET_STORE_BRANCHES",
  },
]
`;

exports[`getStoreBranches should handle failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/'store/undefined/branches",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "ok": false,
          "statusText": "not found",
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStoreBranches should handle failure 2`] = `
Array [
  Object {
    "payload": Object {
      "error": [Error: not found],
    },
    "type": "GET_STORE_BRANCHES_FAIL",
  },
]
`;

exports[`getStoreBranches should handle success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/'store/undefined/branches",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "json": [Function],
          "ok": true,
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStoreBranches should handle success 2`] = `
Array [
  Object {
    "payload": Object {
      "branches": Array [
        Object {
          "id": 2,
          "name": "store 2",
        },
      ],
    },
    "type": "GOT_STORE_BRANCHES",
  },
]
`;
